semantic-core/src/Control/Effect/Readline.hs

```diff
@@ -11,46 +11,40 @@ module Control.Effect.Readline
 , ReadlineC (..)
 , runReadline
 , runReadlineWithHistory
-, TransC (..)
 , ControlIOC (..)
 , runControlIO
 ) where
 
 import Prelude hiding (print)
 
 import Control.Effect.Carrier
+import Control.Effect.Lift
 import Control.Effect.Reader
-import Control.Effect.Sum
-import Control.Monad
 import Control.Monad.IO.Class
 import Control.Monad.Trans.Class
-import Data.Coerce
 import Data.Int
 import Data.String
 import Data.Text.Prettyprint.Doc
 import Data.Text.Prettyprint.Doc.Render.Text
+import GHC.Generics (Generic1)
 import System.Console.Haskeline hiding (Handler, handle)
 import System.Directory
 import System.FilePath
 
 data Readline (m :: * -> *) k
-  = Prompt String (Maybe String -> k)
-  | forall a . Print (Doc a) k
-  | AskLine (Line -> k)
+  = Prompt String (Maybe String -> m k)
+  | Print AnyDoc (m k)
+  | AskLine (Line -> m k)
+  deriving stock (Functor, Generic1)
+  deriving anyclass (Effect, HFunctor)
 
-deriving instance Functor (Readline m)
-
-instance HFunctor Readline where
-  hmap _ = coerce
-
-instance Effect Readline where
-  handle state handler = coerce . fmap (handler . (<$ state))
+newtype AnyDoc = AnyDoc { unAnyDoc :: forall a . Doc a }
```

Comment:
The existential quantifier in `Print` was unnecessary; universal quantification is enough. However, deriving a `Generic1` instance required us to move said quantifier into the above `newtype`.